02_classify_mutations/nbconverted/plot_expression_gene_sets.py

```diff
@@ -0,0 +1,245 @@
+    
+all_results_df.sort_values(by='gene_set', inplace=True)
+all_results_df.sort_values(by='p_value').head(10)
+
+
+# In[7]:
+
+
+all_results_df['nlog10_p'] = -np.log10(all_results_df.corr_pval)
+
+names_map = {
+    'top_50': 'top mutated genes',
+    'random_50': 'random genes',
+    'vogelstein': 'Vogelstein et al.'
+}
+
+all_results_df.gene_set.replace(names_map, inplace=True)
+
+sns.set({'figure.figsize': (24, 6)})
+sns.set_style('whitegrid')
+fig, axarr = plt.subplots(1, 3)
+
+# all plots should have the same axes for a fair comparison
+xlim = (-0.2, 1.0)
+y_max = all_results_df.nlog10_p.max()
+ylim = (0, y_max+3)
+
+# function to add gene labels to points
+def label_points(x, y, gene, ax):
+    text_labels = []
+    a = pd.DataFrame({'x': x, 'y': y, 'gene': gene})
+    for i, point in a.iterrows():
+        if point['y'] > -np.log10(SIG_ALPHA / 1000):
+            text_labels.append(
+                ax.text(point['x'], point['y'], str(point['gene']))
+            )
+    return text_labels
+
+# plot mutation prediction from expression, in a volcano-like plot
+for ix, gene_set in enumerate(sorted(all_results_df.gene_set.unique())):
+    ax = axarr[ix]
+    data_results_df = all_results_df[all_results_df.gene_set == gene_set]
+    sns.scatterplot(data=data_results_df, x='delta_mean', y='nlog10_p', hue='reject_null',
+                    hue_order=[False, True], ax=ax)
+    # add vertical line at 0
+    ax.axvline(x=0, linestyle='--', linewidth=1.25, color='black')
+    # add horizontal line at statistical significance threshold
+    l = ax.axhline(y=-np.log10(SIG_ALPHA), linestyle='--', linewidth=1.25)
+    # label horizontal line with significance threshold
+    # (matplotlib makes this fairly difficult, sadly)
+    ax.text(0.9, -np.log10(SIG_ALPHA)+0.01,
+            r'$\mathbf{{\alpha = {}}}$'.format(SIG_ALPHA),
+            va='center', ha='center', color=l.get_color(),
+            backgroundcolor=ax.get_facecolor())
+    ax.set_xlabel('AUPR(signal) - AUPR(shuffled)')
+    ax.set_ylabel(r'$-\log_{10}($adjusted $p$-value$)$')
+    ax.set_xlim(xlim)
+    ax.set_ylim(ylim)
+    ax.legend(title=r'Reject $H_0$', loc='upper left')
+    ax.set_title(r'Mutation prediction, {}'.format(gene_set))
+
+    # label genes and adjust text to not overlap
+    # automatic alignment isn't perfect, can align by hand in inkscape if necessary
+    text_labels = label_points(data_results_df['delta_mean'],
+                               data_results_df['nlog10_p'],
+                               data_results_df.gene,
+                               ax)
+    
+    adjust_text(text_labels,
+                ax=ax, 
+                expand_text=(1., 1.),
+                lim=5)
+    
+    print('{}: {}/{} ({:.4f})'.format(
+        gene_set,
+        np.count_nonzero(data_results_df.reject_null),
+        data_results_df.shape[0],
+        np.count_nonzero(data_results_df.reject_null) / data_results_df.shape[0]
+    ))
+
+
+# In[8]:
```

Comment:
Is there a reason that the gene distribution in the Vogelstein  is different from the distribution in the others? It feels as though the Vogelstein genes are all predictable, where some small subset of the others are.

Reply:
We were hoping to show that if we select genes based on prior knowledge that they're involved in cancer (the Vogelstein gene set), we can do better (i.e. shift the distribution upward toward higher AUPR values) compared to just selecting genes based on how frequently they're mutated.

So we think/hope that's the reason - just because a gene is frequently mutated in cancer doesn't mean it has any functional effect (they could be passenger mutations, e.g. because the gene is large). And, conversely, lots of strong drivers/highly predictable genes in the Vogelstein gene set aren't that frequently mutated, or are only mutated in a specific subset of samples.